File: sdks/python/docflow/batch_processor.py
"""Batch processor for multi-file queue processing."""

import asyncio
import uuid
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor
from datetime import datetime
from pathlib import Path
from typing import Any, Callable, Dict, List, Optional, Union
from queue import Queue
import threading

from .types import (
    BatchConfig,
    BatchJob,
    JobStatus,
    RAGConfig,
    RAGDocument,
    OutputFormat,
)


class BatchProcessor:
    """Batch processor with queue and parallel processing support.
    
    Supports mixed format inputs (PDF, DOCX, Excel, etc.) in the same batch.
    Provides queue management and parallel processing capabilities.
    
    Example:
        >>> config = RAGConfig(
        ...     llm_processing=[LLMProcessingMode.ALL],
        ...     output_format=OutputFormat.PDF
        ... )
        >>> batch = BatchProcessor(config, max_workers=4)
        >>> 
        >>> # Sync processing
        >>> results = batch.process_files([
        ...     "report.pdf", "data.xlsx", "notes.docx"
        ... ])
        >>> 
        >>> # Async processing with queue
        >>> job_id = await batch.enqueue(files)
        >>> status = await batch.get_status(job_id)
        >>> results = await batch.get_result(job_id)
    """
    
    def __init__(
        self,
        rag_config: Optional[RAGConfig] = None,
        batch_config: Optional[BatchConfig] = None,
        max_workers: int = 4,
    ) -> None:
        """Initialize batch processor.
        
        Args:
            rag_config: RAG configuration for processing.
            batch_config: Batch processing configuration.
            max_workers: Number of parallel workers.
        """
        self.rag_config = rag_config or RAGConfig()
        self.batch_config = batch_config or BatchConfig(max_workers=max_workers)
        self.max_workers = self.batch_config.max_workers
        
        # Job storage
        self._jobs: Dict[str, BatchJob] = {}
        self._job_lock = threading.Lock()
        
        # Queue for async processing
        self._queue: Queue = Queue(maxsize=self.batch_config.queue_size)
        self._executor: Optional[ThreadPoolExecutor] = None
    
    # ============== Sync Processing ==============
    
    def process_files(
        self,
        files: List[Union[str, Path, bytes, tuple]],
        parallel: bool = True,
    ) -> List[RAGDocument]:
        """Process multiple files synchronously.
        
        Args:
            files: List of file paths, bytes, or (bytes, filename) tuples.
            parallel: Use parallel processing.
            
        Returns:
            List of RAGDocument results.
        """
        from .rag import RAGProcessor
        
        processor = RAGProcessor(self.rag_config)
        
        if parallel and len(files) > 1:
            return self._process_parallel(files, processor)
        else:
            return self._process_sequential(files, processor)
    
    def process_to_pdf(
        self,
        files: List[Union[str, Path, bytes, tuple]],
        output_path: Optional[str] = None,
        merge: bool = False,
    ) -> Union[bytes, List[bytes]]:
        """Process files and output as PDF.
        
        Args:
            files: Input files.
            output_path: Optional output path for merged PDF.
            merge: Whether to merge all into one PDF.
            
        Returns:
            PDF bytes or list of PDF bytes.
        """
        # Set output format to PDF
        config = RAGConfig(
            **{**self.rag_config.__dict__, "output_format": OutputFormat.PDF}
        )
        
        from .rag import RAGProcessor
        processor = RAGProcessor(config)
        
        results = self._process_parallel(files, processor)
        
        if merge:
            return self._merge_pdfs([r.pdf_bytes for r in results if r.pdf_bytes])
        else:
            return [r.pdf_bytes for r in results if r.pdf_bytes]
    
    def _process_parallel(
        self,
        files: List[Union[str, Path, bytes, tuple]],
        processor: Any,
    ) -> List[RAGDocument]:
        """Process files in parallel using ThreadPoolExecutor."""
        results = []
        errors = []
        
        with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
            futures = {
                executor.submit(self._process_single, f, processor): f
                for f in files
            }
            
            for future in futures:
                try:
                    result = future.result(timeout=self.batch_config.timeout_per_file)
                    results.append(result)
                except Exception as e:
                    if self.batch_config.fail_fast:
                        raise
                    errors.append((futures[future], str(e)))
        
        return results
    
    def _process_sequential(
        self,
        files: List[Union[str, Path, bytes, tuple]],
        processor: Any,
    ) -> List[RAGDocument]:
        """Process files sequentially."""
        results = []
        
        for f in files:
            try:
                result = self._process_single(f, processor)
                results.append(result)
            except Exception as e:
                if self.batch_config.fail_fast:
                    raise
        
        return results
    
    def _process_single(
        self,
        file_input: Union[str, Path, bytes, tuple],
        processor: Any,
    ) -> RAGDocument:
        """Process a single file."""
        if isinstance(file_input, (str, Path)):
            return processor.process_file(str(file_input))
        elif isinstance(file_input, tuple):
            data, filename = file_input
            return processor.process(data, filename)
        elif isinstance(file_input, bytes):
            return processor.process(file_input, "unknown.bin")
        else:
            raise ValueError(f"Unsupported input type: {type(file_input)}")
    
    # ============== Async Queue Processing ==============
    
    async def enqueue(
        self,
        files: List[Union[str, Path, bytes, tuple]],
    ) -> str:
        """Add files to processing queue.
        
        Args:
            files: Files to process.
            
        Returns:
            Job ID for tracking.
        """
        job_id = str(uuid.uuid4())
        
        job = BatchJob(
            job_id=job_id,
            status=JobStatus.PENDING,
            total_files=len(files),
            created_at=datetime.now().isoformat(),
        )
        
        with self._job_lock:
            self._jobs[job_id] = job
        
        # Start processing in background
        asyncio.create_task(self._process_queue_job(job_id, files))
        
        return job_id
    
    async def get_status(self, job_id: str) -> BatchJob:
        """Get status of a job.
        
        Args:
            job_id: Job ID.
            
        Returns:
            BatchJob with current status.
        """
        with self._job_lock:
            job = self._jobs.get(job_id)
            if not job:
                raise ValueError(f"Job not found: {job_id}")
            return job
    
    async def get_result(self, job_id: str, wait: bool = True) -> List[RAGDocument]:
        """Get results of a completed job.
        
        Args:
            job_id: Job ID.
            wait: Wait for job to complete.
            
        Returns:
            List of RAGDocument results.
        """
        while wait:
            job = await self.get_status(job_id)
            
            if job.status == JobStatus.COMPLETED:
                return job.results
            elif job.status == JobStatus.FAILED:
                raise RuntimeError(f"Job failed: {job.errors}")
            
            await asyncio.sleep(0.5)
        
        job = await self.get_status(job_id)
        return job.results
    
    async def cancel(self, job_id: str) -> bool:
        """Cancel a pending/processing job.
        
        Args:
            job_id: Job ID.
            
        Returns:
            True if cancelled.
        """
        with self._job_lock:
            job = self._jobs.get(job_id)
            if not job:
                return False
            
            if job.status in [JobStatus.PENDING, JobStatus.PROCESSING]:
                job.status = JobStatus.FAILED
                job.errors["cancelled"] = "Job cancelled by user"
                return True
        
        return False
    
    async def _process_queue_job(
        self,
        job_id: str,
        files: List[Union[str, Path, bytes, tuple]],
    ) -> None:
        """Process a queued job."""
        from .rag import RAGProcessor
        
        with self._job_lock:
            job = self._jobs[job_id]
            job.status = JobStatus.PROCESSING
        
        processor = RAGProcessor(self.rag_config)
        loop = asyncio.get_event_loop()
        
        try:
            with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
                futures = []
                
                for f in files:
                    future = loop.run_in_executor(
                        executor,
                        self._process_single,
                        f,
                        processor,
                    )
                    futures.append((f, future))
                
                for file_input, future in futures:
                    try:
                        result = await asyncio.wait_for(
                            future,
                            timeout=self.batch_config.timeout_per_file,
                        )
                        
                        with self._job_lock:
                            job.results.append(result)
                            job.processed_files += 1
                    
                    except Exception as e:
                        filename = str(file_input) if isinstance(file_input, (str, Path)) else "unknown"
                        
                        with self._job_lock:
                            job.errors[filename] = str(e)
                            job.failed_files += 1
                        
                        if self.batch_config.fail_fast:
                            raise
            
            with self._job_lock:
                job.status = JobStatus.COMPLETED
                job.completed_at = datetime.now().isoformat()
        
        except Exception as e:
            with self._job_lock:
                job.status = JobStatus.FAILED
                job.errors["fatal"] = str(e)
                job.completed_at = datetime.now().isoformat()
    
    # ============== PDF Merge ==============
    
    def _merge_pdfs(self, pdf_list: List[bytes]) -> bytes:
        """Merge multiple PDFs into one."""
        try:
            from pypdf import PdfMerger
        except ImportError:
            try:
                from PyPDF2 import PdfMerger
            except ImportError:
                raise ImportError("pypdf or PyPDF2 is required for PDF merging")
        
        from io import BytesIO
        
        merger = PdfMerger()
        
        for pdf_bytes in pdf_list:
            if pdf_bytes:
                merger.append(BytesIO(pdf_bytes))
        
        output = BytesIO()
        merger.write(output)
        merger.close()
        
        return output.getvalue()
    
    # ============== Utility Methods ==============
    
    def list_jobs(self) -> List[BatchJob]:
        """List all jobs."""
        with self._job_lock:
            return list(self._jobs.values())
    
    def clear_completed(self) -> int:
        """Clear completed jobs from memory.
        
        Returns:
            Number of jobs cleared.
        """
        with self._job_lock:
            completed = [
                jid for jid, job in self._jobs.items()
                if job.status in [JobStatus.COMPLETED, JobStatus.FAILED]
            ]
            
            for jid in completed:
                del self._jobs[jid]
            
            return len(completed)

File: sdks/go/docflow/batch_processor.go
package docflow

import (
	"fmt"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/xgaslan/docflow/sdks/go/docflow/config"
	"github.com/xgaslan/docflow/sdks/go/docflow/rag"
)

// jobRequest wraps a job and its files for processing
type jobRequest struct {
	job   *BatchJob
	files []string
}

// BatchProcessor handles multi-file queue processing.
type BatchProcessor struct {
	ragConfig   config.RAGConfig
	batchConfig config.BatchConfig
	maxWorkers  int

	jobs    sync.Map // map[string]*BatchJob
	queue   chan jobRequest
	workers sync.WaitGroup
}

// NewBatchProcessor creates a new batch processor.
func NewBatchProcessor(ragConfig config.RAGConfig, batchConfig config.BatchConfig) *BatchProcessor {
	if batchConfig.MaxWorkers <= 0 {
		batchConfig.MaxWorkers = 4
	}
	if batchConfig.QueueSize <= 0 {
		batchConfig.QueueSize = 100
	}

	bp := &BatchProcessor{
		ragConfig:   ragConfig,
		batchConfig: batchConfig,
		maxWorkers:  batchConfig.MaxWorkers,
		queue:       make(chan jobRequest, batchConfig.QueueSize),
	}

	bp.startWorkers()
	return bp
}

func (bp *BatchProcessor) startWorkers() {
	for i := 0; i < bp.maxWorkers; i++ {
		go bp.worker()
	}
}

func (bp *BatchProcessor) worker() {
	for req := range bp.queue {
		bp.processJob(req)
	}
}

func (bp *BatchProcessor) processJob(req jobRequest) {
	job := req.job
	job.Status = config.JobStatusProcessing
	processor := rag.NewRAGProcessor(bp.ragConfig)

	// In a real implementation, we would process files in parallel too if job has multiple files
	// For simplicity here, we process files sequentially or we could spawn goroutines
	// But since we have worker pool for jobs, maybe files within job?
	// The Python implementation uses thread pool for files.
	// We can do semantic equivalent using semaphores or separate worker pool for files if needed.
	// For now, let's just loop.

	// Assuming job.Files is somehow passed or stored.
	// The Python enqueue takes 'files' list. BatchJob struct might need to store file paths.
	// Let's assume BatchJob has a way to access files (not defined in types.go yet? Check types.go)

	// Wait, types.go BatchJob struct:
	// type BatchJob struct {
	// 	JobID          string                  `json:"job_id"`
	// 	Status         JobStatus               `json:"status"`
	// 	Results        []rag.RAGDocument       `json:"results,omitempty"`
	// 	Errors         map[string]string       `json:"errors,omitempty"`
	// 	TotalFiles     int                     `json:"total_files"`
	// 	ProcessedFiles int                     `json:"processed_files"`
	// 	FailedFiles    int                     `json:"failed_files"`
	// 	CreatedAt      time.Time               `json:"created_at"`
	// 	CompletedAt    time.Time               `json:"completed_at,omitempty"`
	// }
	// It doesn't store input files! Python stores them in closure or somewhere?
	// Python: enqueue(files) -> creates job -> starts _process_queue_job(job_id, files).
	// So the job struct itself doesn't strictly need to hold the file list if the processor function closes over it.
	// However, in Go, we need to pass the file list to the worker via the channel struct.
	// So we should define a wrapper or extend BatchJob.

	// Process files
	for _, f := range req.files {
		doc, err := processor.ProcessFile(f)
		if err != nil {
			job.FailedFiles++
			if job.Errors == nil {
				job.Errors = make(map[string]string)
			}
			job.Errors[f] = err.Error()

			if bp.batchConfig.FailFast {
				job.Status = config.JobStatusFailed
				return
			}
			continue
		}

		job.Results = append(job.Results, *doc)
		job.ProcessedFiles++
	}

	if job.Status != config.JobStatusFailed {
		job.Status = config.JobStatusCompleted
		now := time.Now()
		job.CompletedAt = &now // Assumes CreatedAt/CompletedAt are *time.Time or interface{}
	}
}

// Enqueue adds files to the processing queue.
func (bp *BatchProcessor) Enqueue(files []string) (string, error) {
	jobID := uuid.New().String()
	now := time.Now()
	job := &BatchJob{
		JobID:      jobID,
		Status:     config.JobStatusPending,
		TotalFiles: len(files),
		CreatedAt:  &now,
		Errors:     make(map[string]string),
	}

	bp.jobs.Store(jobID, job)

	// In a real app we might want to persist this.

	// Send to worker (this effectively limits concurrency by queue size/workers)
	// Only if I change channel type.
	// Let's define queue type properly.
	select {
	case bp.queue <- jobRequest{job: job, files: files}:
		return jobID, nil
	default:
		return "", fmt.Errorf("queue is full")
	}
}

// ProcessFiles synchronously processes files.
func (bp *BatchProcessor) ProcessFiles(files []string) ([]*rag.RAGDocument, error) {
	processor := rag.NewRAGProcessor(bp.ragConfig)
	var results []*rag.RAGDocument

	for _, f := range files {
		doc, err := processor.ProcessFile(f)
		if err != nil {
			if bp.batchConfig.FailFast {
				return nil, err
			}
			continue
		}
		results = append(results, doc)
	}
	return results, nil
}

// GetStatus returns the status of a job.
func (bp *BatchProcessor) GetStatus(jobID string) (*BatchJob, error) {
	if val, ok := bp.jobs.Load(jobID); ok {
		return val.(*BatchJob), nil
	}
	return nil, fmt.Errorf("job not found: %s", jobID)
}

// GetResult returns the results of a completed job.
func (bp *BatchProcessor) GetResult(jobID string) ([]rag.RAGDocument, error) {
	job, err := bp.GetStatus(jobID)
	if err != nil {
		return nil, err
	}
	if job.Status == config.JobStatusFailed {
		return nil, fmt.Errorf("job failed")
	}
	return job.Results, nil
}
